File: Competitve Programming/Australian Informatics Olympiad/2018/Janitor/main.py
#!/usr/bin/env python


infile = open("janitorin.txt")
outfile = open("janitorout.txt", "w")

###Infile#####################################
data = [int(i) for i in infile.read().split()]
R, C, Q = data[0:3]

h = [[0 for j in range(C)] for i in range(R)]
for i in range(R):
    for j in range(C):
        h[i][j] = data[3+i*C + j]
#print(region)
#print(h)
#############################################
def output():
    #print(tilesToFill)
    outfile.write(str(tilesToFill)+"\n")

####################################################
directions = [[0,1],[0,-1],[1,0],[-1,0]]
filling = [[False for j in range(C)] for i in range(R)]
tilesToFill = 0 #No tiles to put water on
def checktile(x, y):
    toFill = True #Assume need put water
    #Neighbours
    for dx, dy in directions:
        nx, ny = x+dx, y+dy
        if (nx < 0 or ny < 0 or nx >= C or ny >=R):
            continue
        #print(x, y, nx, ny, h[ny][nx])
        if h[ny][nx] > h[y][x]:
            toFill = False #Should let water flow on it
            break
    if (filling[y][x] != toFill):
        filling[y][x] = toFill
        global tilesToFill
        if toFill:tilesToFill += 1
        #Remove if like no need to be filled anymore. For other round of filling
        else: tilesToFill-=1 
for i in range(R):
    for j in range(C):
        checktile(j, i)
output()

replacements = [int(i) for i in data[3+R*C:3+R*C + Q*3]]
for i in range(Q):
    ri = replacements[i*3]-1
    ci = replacements[i*3+1]-1
    hi = replacements[i*3+2]
    #print(ri, ci, hi)
    h[ri][ci] = hi
    x, y = ci, ri
    checktile(x, y)
    for dx, dy in directions:
        nx, ny = x+dx, y+dy
        if (nx < 0 or ny < 0 or nx >= C or ny >=R):
            continue
        else: checktile(nx, ny)
    output()

infile.close()
outfile.close()

File: Competitve Programming/Australian Informatics Olympiad/2018/Janitor/main.cpp
#include <iostream>
using namespace std;
int r, c, q, s[11][100010], Next[4][2] = {{1,0},{-1,0},{0,1},{0,-1}}, sum, cx, cy, ch;

//Whether location has been filled with water
bool filled[11][100010];

void checkTile(int x, int y){
    int nx, ny; //Neighbouring coordinates
    
    //Assume is higher, should put water on it
    bool flag = true;
    //For neighbours
    for(int i = 0; i <= 3; i++){
        //Set next coordinates
        nx = x + Next[i][0], ny = y + Next[i][1];
        //Ignore if out of range
        if(nx < 1 || nx > r || ny < 1 || ny > c)continue;
        //If neighbour is higher, 
        if(s[nx][ny] > s[x][y]){flag = false; break;}
    }
    //Set pos to appropriate val
    if(filled[x][y] != flag){
        filled[x][y] = flag;
        //If its to be filled,  
        if(flag) sum++;
        //If its not to be filled
        else sum--;
    }
}
int main(){
    freopen("janitorin.txt", "r", stdin);
    freopen("janitorout.txt", "w", stdout);
    scanf("%d%d%d", &r, &c, &q);
    
    //Get Layout of map
    for(int i = 1; i <= r; i++)
        for(int j = 1; j <= c; j++)
            scanf("%d", &s[i][j]);

    //In map
    for(int i = 1; i <= r; i++)
        for(int j = 1; j <= c; j++)
            checkTile(i, j);
    //Number of tiles filled
    printf("%d\n", sum);

    for(int l = 1; l <= q; l++){
        scanf("%d%d%d", &cx, &cy, &ch);
        s[cx][cy] = ch;
        checkTile(cx, cy);
        int nx, ny;
        for(int i = 0; i <= 3; i++){
            nx = cx + Next[i][0], ny = cy + Next[i][1];
            if(nx < 1 || nx > r || ny < 1 || ny > c)continue;
            checkTile(nx, ny);
        }
        printf("%d\n", sum);
    }
    return 0;
}
